tf2_ros/include/tf2_ros/transform_listener.h

```diff
@@ -29,65 +29,102 @@
 
-namespace tf2_ros{
+
+namespace tf2_ros
+{
 
 /** \brief This class provides an easy way to request and receive coordinate frame transform information.
  */
 class TransformListener
 {
-
 public:
   /**@brief Constructor for transform listener */
   TF2_ROS_PUBLIC
-  TransformListener(tf2::BufferCore& buffer, bool spin_thread = true);
-  
-  TF2_ROS_PUBLIC
-  TransformListener(tf2::BufferCore& buffer, rclcpp::Node::SharedPtr nh, bool spin_thread = true);
+  explicit TransformListener(tf2::BufferCore & buffer, bool spin_thread = true);
+
+  template<class NodeT, class AllocatorT = std::allocator<void>>
+  TransformListener(
+    tf2::BufferCore & buffer,
+    NodeT && node,
+    bool spin_thread = true,
+    const rclcpp::QoS & qos = rclcpp::QoS(100),
+    const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> & options =
+      rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>())
+  : buffer_(buffer)
+  {
+    init(node, spin_thread, qos, options);
+  }
 
   TF2_ROS_PUBLIC
-  ~TransformListener();
+  virtual ~TransformListener();
 
 private:
+  template<class NodeT, class AllocatorT = std::allocator<void>>
+  void init(
+    NodeT && node,
+    bool spin_thread,
+    const rclcpp::QoS & qos = rclcpp::QoS(100),
+    const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> & options =
+      rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>())
+  {
+    using callback_t = std::function<void(tf2_msgs::msg::TFMessage::SharedPtr)>;
+    callback_t cb = std::bind(&TransformListener::subscription_callback, this, std::placeholders::_1, false);
+    callback_t static_cb = std::bind(&TransformListener::subscription_callback, this, std::placeholders::_1, true);
 
-  /// Initialize this transform listener, subscribing, advertising services, etc.
-  void init();
-  void initThread();
+    message_subscription_tf_ = rclcpp::create_subscription<tf2_msgs::msg::TFMessage>(
+      std::forward<NodeT>(node),
+      "/tf",
+      qos,
+      std::move(cb),
+      options);
+    message_subscription_tf_static_ = rclcpp::create_subscription<tf2_msgs::msg::TFMessage>(
+      std::forward<NodeT>(node),
+      "/tf_static",
+      qos,
+      std::move(static_cb),
+      options);
+
+    if (spin_thread) {
+      initThread(node->get_node_base_interface());
+    }
+  }
+
+  void initThread(rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base_interface);
 
   /// Callback function for ros message subscriptoin
-  void subscription_callback(const tf2_msgs::msg::TFMessage::SharedPtr msg);
-  void static_subscription_callback(const tf2_msgs::msg::TFMessage::SharedPtr msg);
-  void subscription_callback_impl(const tf2_msgs::msg::TFMessage::SharedPtr msg, bool is_static);
+  void subscription_callback(tf2_msgs::msg::TFMessage::SharedPtr msg, bool is_static);
 
   // ros::CallbackQueue tf_message_callback_queue_;
-  std::thread* dedicated_listener_thread_;
-  rclcpp::Node::SharedPtr node_;
+  std::thread * dedicated_listener_thread_ = nullptr;
```

Comment:
Can we get away with `std::unique_ptr` here?

Reply:
I have thought about this and would like to keep this into a separate PR.

Reply:
Sounds good to me! I'd be happy to file it if you don't have time. Let me know

Reply:
please go ahead and file it. It really shouldn't be more than just a unique_ptr with a custom deleter (basically what's done in the destructor for now).